### PR TITLE
[WIP] gh-93744: Remove configure --with-cxx-main option

### PR DESCRIPTION
**Pull request by vstinner**
Remove "configure --with-cxx-main" option: it didn't work for many
years.  Remove the following configure and Makefile variables: CXX,
LDCXXSHARED, MAINCC.

* MAINCC was added by issue gh-42471:
  commit 0f48d98b740110a672b62d467af192ec160e56ba.
  Previously, --with-cxx-main was named --with-cxx.
* LDCXXSHARED was added by issue gh-42093:
  commit 0000295fe3b316f78732d543e15ae69efe168bed.
* CXX variable was used by MAINCC and LDCXXSHARED variables.

### CODE REVIEW

**Comment by tiran** [comment]
The Makefile values are included in sysconfig data file and are available through ``sysconfig``. Are you sure that no 3rd party project relies on the settings?

File: configure.ac [hunk 1]
```diff
@@ -807,62 +804,6 @@ rm -f conftest.c conftest.out
 # _POSIX_SOURCE, _POSIX_1_SOURCE, and more
 AC_USE_SYSTEM_EXTENSIONS
 
-AC_SUBST(CXX)
```
Comment (tiran) [inline configure.ac:810]: ``makesetup`` and final ``Makefile`` rely on ``CXX`` variable.

**Comment by vstinner** [comment]
It seems like removing CXX is dangerous: it seems to be used, so it would be safer... to just keep it, just in case of doubt :-) MAINCC and LDCXXSHARED don't seem to be used outside Python itself.

> Are you sure that no 3rd party project relies on the settings?

I ran a code search in the pypa GitHub organization:

* https://github.com/search?q=org%3Apypa+MAINCC : MAINCC is only mentioned in one issue, in a dump of sysconfig variables. It's not used in the code.
* https://github.com/search?q=org%3Apypa+LDCXXSHARED : LDCXXSHARED is only mentioned in one issue, in a dump of sysconfig variables. It's not used in the code.
* https://github.com/search?q=org%3Apypa+CXX : CXX is used in the code, but it seems like it's more about the CXX environment variable. But there is a variant of sysconfig.customize_compiler() which uses the sysconfig environment variable.

> makesetup and final Makefile rely on CXX variable.

Yep, as I wrote.

makesetup:

```
			*.cc)  obj=`basename $src .cc`.o; cc='$(CXX)';;
			*.c++) obj=`basename $src .c++`.o; cc='$(CXX)';;
			*.C)   obj=`basename $src .C`.o; cc='$(CXX)';;
			*.cxx) obj=`basename $src .cxx`.o; cc='$(CXX)';;
			*.cpp) obj=`basename $src .cpp`.o; cc='$(CXX)';;
```

Python Git repository has no file with extensions: .cc, .c++, .C, .cxx.

There are a few files with .cpp extensions:

```
Lib/test/_testcppext.cpp
PC/pyshellext.cpp
PC/python_uwp.cpp
Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp
Tools/msi/bundle/bootstrap/pch.cpp
Tools/msi/bundle/bootstrap/pythonba.cpp
```

* Lib/test/_testcppext.cpp is built with setuptools in test_cppext
* PC/pyshellext.cpp and PC/python_uwp.cpp: I don't think that the Windows build system (VS) uses makesetup.
* Tools/msi/bundle/: I don't expect it to use makesetup.

I only see "CXX" at one line in the final Makefile: "CXX='$(CXX)'", does it copy the environment variable or the Makefile variable?

```
buildbottest: all
		-@if which pybuildbot.identify >/dev/null 2>&1; then \
			pybuildbot.identify "CC='$(CC)'" "CXX='$(CXX)'"; \
		fi
		$(TESTRUNNER) -j 1 -u all -W --slowest --fail-env-changed --timeout=$(TESTTIMEOUT) $(TESTOPTS)
```

**Comment by erlend-aasland** [comment]
> I only see "CXX" at one line in the final Makefile: "CXX='$(CXX)'", does it copy the environment variable or the Makefile variable?

It copies the Makefile variable `CXX` to an environment variable.

BTW, what did you conclude here? The cxx stuff stays?

**Comment by vstinner** [comment]
> BTW, what did you conclude here?

I'm not brave. I abandon this PR and I wrote PR #95651 instead which only removes the MAINCC variables but keeps unused CXX and LDCXXSHARED variables.

**Pull request closed without merging**